### PR TITLE
PERF: vectorize _interp_limit

### PR DESCRIPTION
**Pull request by TomAugspurger**
xref https://github.com/pandas-dev/pandas/issues/16584

Here are the timings vs. master

```
       before           after         ratio
     [ff0d1f4d]       [6b1552f1]
-      12.3±0.3ms      2.28±0.04ms     0.18  frame_methods.Interpolate.time_interpolate_some_good_infer
-      12.6±0.7ms      1.15±0.01ms     0.09  frame_methods.Interpolate.time_interpolate_some_good
-           1.30s         53.8±1ms     0.04  frame_methods.Interpolate.time_interpolate

```

I still need to time vs 0.20.1 (will post those timings later, have to run for a bit now). I may have one more spot to optimize.

### CODE REVIEW

**Comment by TomAugspurger** [comment]
Also, the master and 0.20.1 versions scale poorly with `limit`, but that's not benchmarked anywhere. Here's some timings

limit / version | 0.20.1  | master | interp-perf
--------------- | ------- | ------ | -----------
None            | 42.2 ms | 1.37 s | 43.8 ms
50              | 1.45 s  | 1.33 s | 66.2 ms
500             | 1.40  s  | 1.36 s | 62.5 ms

So we're slightly slower than 0.20.1 for the default, but much faster with any limit (notice the units).

I'll add a benchmark for `limit` and `direction`. Anyone know if there's a way to run new benchmarks against old commits?

**Comment by TomAugspurger** [comment]
Just finished vs. 0.20.1, and it's slightly slower but below the 1.1 threshold

```
· Running 6 total benchmarks (2 commits * 1 environments * 3 benchmarks)
[  0.00%] · For pandas commit hash b2bbd1bf:
[  0.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...........................................................................................
[  0.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 16.67%] ··· Running frame_methods.Interpolate.time_interpolate                                                                                  44.8±0.3ms
[ 33.33%] ··· Running frame_methods.Interpolate.time_interpolate_some_good                                                                       1.09±0.03ms
[ 50.00%] ··· Running frame_methods.Interpolate.time_interpolate_some_good_infer                                                                 2.41±0.01ms
[ 50.00%] · For pandas commit hash 20baf972:
[ 50.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...
[ 50.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 66.67%] ··· Running frame_methods.Interpolate.time_interpolate                                                                                  46.7±0.9ms
[ 83.33%] ··· Running frame_methods.Interpolate.time_interpolate_some_good                                                                       1.08±0.02ms
[100.00%] ··· Running frame_methods.Interpolate.time_interpolate_some_good_infer                                                                 2.44±0.07ms
BENCHMARKS NOT SIGNIFICANTLY CHANGED.
```

File: pandas/core/missing.py [hunk 1]
```diff
@@ -630,3 +632,65 @@ def fill_zeros(result, x, y, name, fill):
             result = result.reshape(shape)
 
     return result
+
+
+def _interp_limit0(invalid, fw_limit, bw_limit):
+    "Get idx of values that won't be filled b/c they exceed the limits."
+    for x in np.where(invalid)[0]:
+        if invalid[max(0, x - fw_limit):x + bw_limit + 1].all():
```
Comment (jreback) [inline pandas/core/missing.py]: should rewrite this as well

**Comment by TomAugspurger** [comment]
Oh sorry that's not used anymore. Just left it around for testing that my new implementation was the same. I'll remove it.

File: pandas/core/missing.py [hunk 2]
```diff
@@ -630,3 +632,58 @@ def fill_zeros(result, x, y, name, fill):
...
+        # then we don't even need to care about backwards, just use forwards
+        return f_idx
+    else:
+        b_idx = set(N - 1 - np.asarray(list(inner(invalid[::-1], bw_limit))))
+        if fw_limit == 0:
+            return b_idx
+    return f_idx & b_idx
+
+
+def rolling_window(a, window):
```
Comment (jreback) [inline pandas/core/missing.py]: maybe move this inside _interp_limit to avoid confusion that this is public and/or used externally

**Review by jreback — APPROVED** [review]
this all looks fine
i suspect a pure cython impl might be easier to read though (maybe even faster)

can tackle later if interest

**Comment by TomAugspurger** [comment]
> i suspect a pure cython impl might be easier to read though (maybe even faster)

Yeah, it's kind of unreadable, and I have no idea what that striding stuff is doing.

I'll leave this open if you want to take a look. I'll tag 0.20.2 early tomorrow morning.
We can also do the cython for 0.21

**Comment by jreback** [comment]
no this is all fine maybe make an issue for 0.21 to have a look

**Pull request merged**